base_control/Motor.cpp

```diff
@@ -1,186 +1,186 @@
-}
-
-//////////////////////////////////Get Methods//////////////////////////////////////
-int Motor::getPidTicks() {
-  return pid_ticks_;
-}
-
-int Motor::getOdomTicks() {
-  return odom_ticks_;
-}
-
-double Motor::getLastTicks() {
-  return last_ticks_;
-}
-
-double Motor::getCurrentSpeed() {
-  return current_speed_;
-}
-
-MotorState Motor::getCurrentState() {
-  return current_state_;
-}
\ No newline at end of file
+#include <math.h>
+#include "Encoder.h"
+#include "Motor.h"
+
+//////////////////////////////////Constructor//////////////////////////////////////
+Motor::Motor() {}
+Motor::Motor(const MotorId id, const uint8_t digital_one, const uint8_t digital_two, 
+const uint8_t analog_one, const uint8_t encoder_one, const uint8_t encoder_two) : 
+pid_(kP, kI, kD, kPidMinOutputLimit, kPidMaxOutputLimit, kPidMaxErrorSum, kPidMotorTimeSample) {
+
+  id_ = id;
+  digital_one_ = digital_one;
+  digital_two_ = digital_two;
+  analog_one_ = analog_one;
+  encoder_one_ = encoder_one;
+  encoder_two_ = encoder_two;
+
+  stop();
+  defineOutput();
+
+}
+
+
+//////////////////////////////////Initialization//////////////////////////////////////
+void Motor::defineOutput() {
+  pinMode(digital_one_, OUTPUT);
+  pinMode(digital_two_, OUTPUT);
+  pinMode(analog_one_, OUTPUT);
+  pinMode(encoder_one_, INPUT);
+  pinMode(encoder_two_, INPUT);
+}
+
+void Motor::initEncoders() {
+  switch (id_) {
+    case MotorId::BackLeft:
+      attachInterrupt(digitalPinToInterrupt(encoder_one_), Encoder::backLeftEncoder, RISING);
+    break;
+    case MotorId::FrontLeft:
+      attachInterrupt(digitalPinToInterrupt(encoder_one_), Encoder::frontLeftEncoder, RISING);
+    break;
+    case MotorId::BackRight:
+      attachInterrupt(digitalPinToInterrupt(encoder_one_), Encoder::backRightEncoder, RISING);
+    break;
+    case MotorId::FrontRight:
+      attachInterrupt(digitalPinToInterrupt(encoder_one_), Encoder::frontRightEncoder, RISING);
+    break;
+  }
+}
+
+//////////////////////////////////Motor State//////////////////////////////////////
+void Motor::forward() {
+  
+  analogWrite(analog_one_, pwm_);
+
+  if(current_state_ == MotorState::Forward) {
+    return;
+  }
+  
+  digitalWrite(digital_one_, HIGH);
+  digitalWrite(digital_two_, LOW);
+
+  if(current_state_ != MotorState::Forward) {
+    pid_.reset();
+  }
+  
+  current_state_ = MotorState::Forward;
+}
+
+void Motor::backward() {
+
+  analogWrite(analog_one_, pwm_);
+  
+  if(current_state_ == MotorState::Backward) {
+    return;
+  }
+  
+  digitalWrite(digital_one_, LOW);
+  digitalWrite(digital_two_, HIGH);
+  
+  if(current_state_ != MotorState::Backward) {
+    pid_.reset();
+  }
+  
+  current_state_ = MotorState::Backward;
+}
+
+void Motor::stop() {
+  analogWrite(analog_one_, LOW);
+
+  if(current_state_ == MotorState::Stop) {
+    return;
+  }
+
+  digitalWrite(digital_one_, LOW);
+  digitalWrite(digital_two_, LOW);
+  
+  if(current_state_ != MotorState::Stop) {
```

Comment:
The idea of those ifs conditions was defined here: https://github.com/RoBorregos/Robocup-Home/pull/23/commits/dbc564b574c613a99d4e305199a4a76f3e6edfb7#r407221040

Reply:
Yes, the first set of ifs are great; I'm talking about the second set of ifs. If you pay close attention, you return immediately if you're already on the desired state, and only otherwise you continue with the function. Hence it is redundant to check once more the state, because it hasn't changed since the last check.